### ISSUE TITLE
Loss calculation on discriminator for WGAN and WGAN GP

### ISSUE THREAD

**Issue by Chechgm** [eriklindernoren/Keras-GAN#140]
In the WGAN GP the loss is calculated:
```
# Sample generator input
noise = np.random.normal(0, 1, (batch_size, self.latent_dim))
# Train the critic
d_loss = self.critic_model.train_on_batch([imgs, noise], [valid, fake, dummy])
```
I am wondering whether this is an intended behavior or instead it should be calculated with generated samples from the generator using the "noise" variable.

Thanks!

**Comment by eriklindernoren** [comment]
Yes, that is intended. If you look in `WGANGP.__init__` you can see how the computational graph of the critic is compiled. The critic takes a images sampled from the dataset and noise as input (which will be used by the generator to produce fake samples), and tries to discriminate real from fake samples. The real and fake samples are also used to compute the gradient penalty described in the WGAN-GP paper.

```python
        #-------------------------------
        # Construct Computational Graph
        #       for the Critic
        #-------------------------------

        # Freeze generator's layers while training critic
        self.generator.trainable = False

        # Image input (real sample)
        real_img = Input(shape=self.img_shape)

        # Noise input
        z_disc = Input(shape=(self.latent_dim,))
        # Generate image based of noise (fake sample)
        fake_img = self.generator(z_disc)

        # Discriminator determines validity of the real and fake images
        fake = self.critic(fake_img)
        valid = self.critic(real_img)

        # Construct weighted average between real and fake images
        interpolated_img = RandomWeightedAverage()([real_img, fake_img])
        # Determine validity of weighted sample
        validity_interpolated = self.critic(interpolated_img)

        # Use Python partial to provide loss function with additional
        # 'averaged_samples' argument
        partial_gp_loss = partial(self.gradient_penalty_loss,
                          averaged_samples=interpolated_img)
        partial_gp_loss.__name__ = 'gradient_penalty' # Keras requires function names

        self.critic_model = Model(inputs=[real_img, z_disc],
                            outputs=[valid, fake, validity_interpolated])
        self.critic_model.compile(loss=[self.wasserstein_loss,
                                              self.wasserstein_loss,
                                              partial_gp_loss],
                                        optimizer=optimizer,
                                        loss_weights=[1, 1, 10])
```
